### PR TITLE
[GEOS-10454] Make LanguageUrl mangler using rawKvp instead of HttpServletRequest params

### PR DESCRIPTION
**Pull request by fostash**
[GEOS-10454](https://osgeo-org.atlassian.net/browse/GEOS-10454)

[GEOS-10454] use raw kv params detached from servlet request because of tomcat request lifecycle that cause the reset of a finished request

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by taba90** [comment]
looks good to me. Could you only replace the current commit message using the titile instead of the description from the Jira ticket? thanks!

File: src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java [hunk 1]
```diff
@@ -20,24 +19,23 @@ public void mangleURL(
...
-        String acceptLanguagesParameter = null;
+        final Map<String, Object> rawKvp =
+                Optional.ofNullable(Dispatcher.REQUEST.get().rawKvp).orElse(new HashMap<>());
 
-        HttpServletRequest httpRequest = Dispatcher.REQUEST.get().getHttpRequest();
-        Enumeration<String> parameterNames = httpRequest.getParameterNames();
+        final Optional<String> languageParameter =
+                rawKvp.keySet().stream()
+                        .filter(param -> param.equalsIgnoreCase(LANGUAGE))
+                        .findFirst();
```
Comment (taba90) [inline src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:28]: the raw kvp map is an HasMap implementation that is already ignore cased https://github.com/geoserver/geoserver/blob/9e40cbbccb9626b8d08516c8eacaf5a8f168189b/src/ows/src/main/java/org/geoserver/ows/util/KvpMap.java#L19. We can directly use the contains method of it without iterating before.

File: src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java [hunk 2]
```diff
@@ -20,24 +19,23 @@ public void mangleURL(
...
 
-        for (String parameter : Collections.list(parameterNames)) {
-            if (parameter.equalsIgnoreCase(LANGUAGE)) {
-                languageParameter = parameter;
-            } else if (parameter.equalsIgnoreCase(ACCEPT_LANGUAGES)) {
-                acceptLanguagesParameter = parameter;
-            }
-        }
+        final Optional<String> acceptLanguagesParameter =
```
Comment (taba90) [inline src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:30]: same as before

**Review by taba90 — CHANGES_REQUESTED** [review]
Looks good only a minor change requested.

File: src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java [hunk 3]
```diff
@@ -20,24 +19,13 @@ public void mangleURL(
...
-            }
-        }
+        final Map<String, Object> rawKvp =
+                Optional.ofNullable(Dispatcher.REQUEST.get().rawKvp).orElse(new KvpMap<>());
 
         String language = "";
 
-        if (acceptLanguagesParameter != null) {
-            String acceptLanguages = httpRequest.getParameter(acceptLanguagesParameter);
+        if (rawKvp.containsKey(ACCEPT_LANGUAGES)) {
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:27]: I think this if is redundant.

File: src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java [hunk 4]
```diff
@@ -20,24 +19,13 @@ public void mangleURL(
...
+            String acceptLanguages = (String) rawKvp.get(ACCEPT_LANGUAGES);
             if (acceptLanguages != null) {
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:29]: I would check here if the string is not empty, or even if the string as a comma.

**Comment by nmco** [comment]
Thank you @fostash, looks good to me, I will merge when the build is green.

File: src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java [hunk 5]
```diff
@@ -4,57 +4,72 @@
...
-        }
-
-        if (languageParameter != null) {
-            language = httpRequest.getParameter(languageParameter);
-        }
-
-        if (language != null && language.length() > 0) {
-            kvp.put("Language", language.trim());
-        }
+        Optional.ofNullable(type)
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:32]: Careful about this proliferation of Optional calls, as they come with object allocations, one for the main Optional, plus others for specific methods (e.g. map, flatmap).

In this code, it's probably not harmful, a single request processing normally builds up to a few tens of URLs, so overall the code change introduces the allocation of the order of 100 (ish) new objects that weren't there in the old code. In the economy of a OGC request, that's peanuts.

However, always consider the context in which the code is running, if it's called many times (hundreds of thousands, millions, not unusual when dealing with GIS data) best to stay away from optionals and streams due to their memory and execution costs.

Comment (fostash) [reply src/main/src/main/java/org/geoserver/ows/LanguageURLMangler.java:32]: I perfectly got your point, and here the `type` parameter wrapped in an Optional is for sure useless. I will revert it with the if.

**Comment by aaime** [comment]
Legit formatting check failure in the QA build:

```
Error: 7,275 [ERROR] Failed to execute goal com.diffplug.spotless:spotless-maven-plugin:2.20.0:check (default) on project gs-main: The following files had format violations:
Error: 7,275 [ERROR]     src/main/java/org/geoserver/ows/LanguageURLMangler.java
Error: 7,275 [ERROR]         @@ -37,8 +37,8 @@
Error: 7,275 [ERROR]          
Error: 7,275 [ERROR]          ····/**
Error: 7,275 [ERROR]          ·····*·@return·Getting·the·rawKvp·from·the·Request·inside·the·thread·local·search·for·parameter
Error: 7,275 [ERROR]         -·····*·Language·or·AcceptLanguages·and·return·the·first·one·if·present·or·the·first·occurrence
Error: 7,275 [ERROR]         -·····*·of·a·language·code·(if·present)·inside·the·second·one
Error: 7,275 [ERROR]         +·····*·····Language·or·AcceptLanguages·and·return·the·first·one·if·present·or·the·first·occurrence
Error: 7,275 [ERROR]         +·····*·····of·a·language·code·(if·present)·inside·the·second·one
Error: 7,275 [ERROR]          ·····*/
Error: 7,275 [ERROR]          ····protected·Optional<Object>·processLanguageParam()·{
Error: 7,275 [ERROR]          ········return·Optional.ofNullable(Dispatcher.REQUEST.get())
Error: 7,275 [ERROR]         @@ -50,7 +50,7 @@
Error: 7,275 [ERROR]          ····/**
Error: 7,275 [ERROR]          ·····*·@param·rawKvp·Map·of·request·parameter
Error: 7,275 [ERROR]          ·····*·@return·the·value·of·Language·parameter·inside·the·{@link·Request#rawKvp}·map·or·an
Error: 7,275 [ERROR]         -·····*·Optional.empty
Error: 7,275 [ERROR]         +·····*·····Optional.empty
Error: 7,275 [ERROR]          ·····*/
Error: 7,275 [ERROR]          ····protected·Optional<Object>·getLanguage(Map<String,·Object>·rawKvp)·{
Error: 7,275 [ERROR]          ········return·Optional.ofNullable(rawKvp.get(LANGUAGE)).filter(filterEmtpyString);
Error: 7,275 [ERROR] Run 'mvn spotless:apply' to fix these violations.
```

Just run a maven build before committing, it will reformat the code for you.

**Pull request merged**